### ISSUE TITLE
Namespaces stay in forever terminating state

### ISSUE THREAD

**Issue by cooktheryan** [kubernetes-sigs/kubefed#954]
**What would you like to be added**: Sometimes project stay in a forever terminating state if everything isn't deleted before trying to delete a project.

**Why is this needed**: Customers and myself will mess this up and then a project is stuck forever in a terminating state.

/kind feature

**Comment by marun** [comment]
@cooktheryan What do you mean by 'project' - an OpenShift `Project` API resource?  Also, please provide sufficient details to enable reproduction of the problem.

**Comment by cooktheryan** [comment]
@marun when I say project I just mean a namespace. I'll get you steps to reproduce in the morning.

**Comment by cooktheryan** [comment]
@marun so here is the run down of how to make it happen.

1) Create namespace
2) federate cluster(s)
3) deploy some sort of federated object (federatedclusterrole)
4) Forget to delete federated object
5) unfederate cluster(s)
6) unsubscribe from operator
7) attempt to delete project

**Comment by marun** [comment]
Apologies for this lengthy response.  

tl;dr; finalizers are the reason you're experiencing this problem, and finalizers are a necessary evil for KubeFed.  We need to both improve the documentation around them and simplify their removal if required.

By 'federate clusters' I'm assuming you mean 'register a cluster with the control plane'?  `Federated` was previously an overloaded term, and the name change from `FederatedCluster` to `KubeFedCluster` was intended to ensure that 'a federated resource' could unambigously be defined as 'a resource that defines configuration for more than one cluster' (which was never true of `FederatedCluster`).

Name pedantry aside, the likely reason the namespace is stuck in terminating state is that one or more contained federated resources still have the `kubefed.k8s.io/sync-controller` finalizer.  The sync controller for a federated type adds this finalizer to every federated resource it encounters.  This ensures that the controller has an opportunity to observe deletion of the resource and perform pre-deletion removal of managed resources in member clusters before the federated resource is removed.  This is [called out in the userguide](https://github.com/kubernetes-sigs/kubefed/blob/master/docs/userguide.md#deletion-policy), though perhaps not in sufficient detail.

In the case of a cluster-scoped control plane, it would be unusual for federated resources to exist in `kube-federation-system` and other namespaces containing federated resources could be removed prior to stopping the control plane.  Removing any of these other namespaces (with the control plane still running) would allow sync controllers the opportunity to perform pre-deletion cleanup and remove the finalizer so that garbage collection could proceed.  

For a namespace-scoped control plane (which I'm assuming is the case you're describing), federated resources would only exist in the `kube-federation-system` namespace.  Deleting the namespace wouldn't reliably result in finalizer removal because the control plane would also be removed and might not get a change to finalize all resources before it was shut down. 

If user intent was to remove resources (in member clusters) managed by federated resources, deletion of all federated resources in the namespace should be completed prior to control plane shutdown.  iirc `kubectl` doesn't support deleting the contents of a namespace without initiating namespace deletion, maybe `kubefedctl` needs to support deleting all federated resources in a namespace?  And maybe we could rely on api groups to enable this (cc: @pmorie)?

If user intent was to retain resources in member clusters, it would be possible to shut down the control plane and then manually remove the finalizer from all resources in the namespace.  This could be done via the cli manually removing the finalizer via `kubectl edit`, but that would be laborious.  `kubectl patch` could be used as follows, but might result in the unintended removal of finalizers set by other controllers (which is why the userguide doesn't mention it):

```
kubectl patch <resource type> <resource name> -n <namespace-name> -p '{"metadata":{"finalizers":null}}
```

Similar to deletion, maybe `kubefedctl` should support finalizer removal from any given resource and also from all resources in a namespace to ensure that resource deletion will work as expected with the control plane shutdown?
